File: sysmayal2/sysmayal2/report/compliance_status_report/compliance_status_report.py
"""
Compliance Status Report

This report provides comprehensive overview of product compliance status
across different countries and regulatory frameworks.
"""

import frappe
from frappe import _
from frappe.utils import date_diff, nowdate

def execute(filters=None):
    """Execute the compliance status report."""
    
    columns = get_columns()
    data = get_data(filters)
    chart_data = get_chart_data(data)
    
    return columns, data, None, chart_data

def get_columns():
    """Define report columns."""
    
    return [
        {
            "fieldname": "product_name",
            "label": _("Product Name"),
            "fieldtype": "Link",
            "options": "Product Compliance",
            "width": 200
        },
        {
            "fieldname": "product_code",
            "label": _("Product Code"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "country",
            "label": _("Country"),
            "fieldtype": "Link", 
            "options": "Country",
            "width": 120
        },
        {
            "fieldname": "compliance_status",
            "label": _("Compliance Status"),
            "fieldtype": "Data",
            "width": 130
        },
        {
            "fieldname": "compliance_percentage",
            "label": _("Compliance %"),
            "fieldtype": "Percent",
            "width": 100
        },
        {
            "fieldname": "risk_level",
            "label": _("Risk Level"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "approval_status", 
            "label": _("Approval Status"),
            "fieldtype": "Data",
            "width": 130
        },
        {
            "fieldname": "testing_status",
            "label": _("Testing Status"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "next_review_date",
            "label": _("Next Review"),
            "fieldtype": "Date",
            "width": 110
        },
        {
            "fieldname": "days_to_review",
            "label": _("Days to Review"),
            "fieldtype": "Int",
            "width": 110
        },
        {
            "fieldname": "expiry_date",
            "label": _("Expiry Date"),
            "fieldtype": "Date",
            "width": 110
        },
        {
            "fieldname": "days_to_expiry",
            "label": _("Days to Expiry"),
            "fieldtype": "Int",
            "width": 110
        },
        {
            "fieldname": "responsible_person",
            "label": _("Responsible Person"),
            "fieldtype": "Link",
            "options": "User",
            "width": 150
        },
        {
            "fieldname": "manufacturer",
            "label": _("Manufacturer"),
            "fieldtype": "Link",
            "options": "Distribution Organization",
            "width": 150
        }
    ]

def get_data(filters):
    """Get report data based on filters."""
    
    conditions = get_conditions(filters)
    
    query = f"""
        SELECT 
            pc.name,
            pc.product_name,
            pc.product_code,
            pc.country,
            pc.compliance_status,
            pc.compliance_percentage,
            pc.risk_level,
            pc.approval_status,
            pc.testing_status,
            pc.next_review_date,
            pc.expiry_date,
            pc.responsible_person,
            pc.manufacturer,
            CASE 
                WHEN pc.next_review_date IS NOT NULL 
                THEN DATEDIFF(pc.next_review_date, CURDATE())
                ELSE NULL 
            END as days_to_review,
            CASE 
                WHEN pc.expiry_date IS NOT NULL 
                THEN DATEDIFF(pc.expiry_date, CURDATE())
                ELSE NULL 
            END as days_to_expiry
        FROM `tabProduct Compliance` pc
        WHERE pc.docstatus < 2
        {conditions}
        ORDER BY 
            pc.country,
            CASE pc.compliance_status
                WHEN 'Non-Compliant' THEN 1
                WHEN 'Expired' THEN 2
                WHEN 'Pending Review' THEN 3
                WHEN 'Partially Compliant' THEN 4
                WHEN 'Compliant' THEN 5
                ELSE 6
            END,
            pc.product_name
    """
    
    return frappe.db.sql(query, as_dict=True)

def get_conditions(filters):
    """Build SQL conditions based on filters."""
    
    conditions = ""
    
    if filters:
        if filters.get("country"):
            conditions += f" AND pc.country = '{filters['country']}'"
            
        if filters.get("compliance_status"):
            conditions += f" AND pc.compliance_status = '{filters['compliance_status']}'"
            
        if filters.get("risk_level"):
            conditions += f" AND pc.risk_level = '{filters['risk_level']}'"
            
        if filters.get("manufacturer"):
            conditions += f" AND pc.manufacturer = '{filters['manufacturer']}'"
            
        if filters.get("responsible_person"):
            conditions += f" AND pc.responsible_person = '{filters['responsible_person']}'"
            
        if filters.get("from_date"):
            conditions += f" AND pc.last_review_date >= '{filters['from_date']}'"
            
        if filters.get("to_date"):
            conditions += f" AND pc.next_review_date <= '{filters['to_date']}'"
            
    return conditions

def get_chart_data(data):
    """Generate chart data for the report."""
    
    # Compliance status distribution
    status_counts = {}
    risk_counts = {}
    country_compliance = {}
    
    for row in data:
        # Status distribution
        status = row.get('compliance_status', 'Unknown')
        status_counts[status] = status_counts.get(status, 0) + 1
        
        # Risk distribution
        risk = row.get('risk_level', 'Unknown')
        risk_counts[risk] = risk_counts.get(risk, 0) + 1
        
        # Country compliance
        country = row.get('country', 'Unknown')
        if country not in country_compliance:
            country_compliance[country] = {'total': 0, 'compliant': 0}
        country_compliance[country]['total'] += 1
        if status == 'Compliant':
            country_compliance[country]['compliant'] += 1
    
    # Calculate compliance rates by country
    country_rates = {}
    for country, data in country_compliance.items():
        rate = (data['compliant'] / data['total'] * 100) if data['total'] > 0 else 0
        country_rates[country] = round(rate, 1)
    
    chart_data = {
        "data": {
            "labels": list(status_counts.keys()),
            "datasets": [{
                "name": "Compliance Status Distribution",
                "values": list(status_counts.values())
            }]
        },
        "type": "donut",
        "height": 300,
        "colors": ["#28a745", "#ffc107", "#dc3545", "#6c757d", "#17a2b8"]
    }
    
    return chart_data

def get_report_summary(data):
    """Generate report summary statistics."""
    
    if not data:
        return {}
    
    total_products = len(data)
    compliant_products = len([d for d in data if d.get('compliance_status') == 'Compliant'])
    non_compliant = len([d for d in data if d.get('compliance_status') == 'Non-Compliant'])
    expired = len([d for d in data if d.get('compliance_status') == 'Expired'])
    
    # Products expiring soon (within 30 days)
    expiring_soon = len([d for d in data if d.get('days_to_expiry') and d['days_to_expiry'] <= 30 and d['days_to_expiry'] > 0])
    
    # Reviews due soon (within 7 days)
    reviews_due = len([d for d in data if d.get('days_to_review') and d['days_to_review'] <= 7 and d['days_to_review'] > 0])
    
    compliance_rate = (compliant_products / total_products * 100) if total_products > 0 else 0
    
    return {
        "total_products": total_products,
        "compliant_products": compliant_products,
        "compliance_rate": round(compliance_rate, 1),
        "non_compliant": non_compliant,
        "expired": expired,
        "expiring_soon": expiring_soon,
        "reviews_due": reviews_due
    }

@frappe.whitelist()
def get_compliance_filters():
    """Get available filter options for the report."""
    
    countries = frappe.get_all("Country", fields=["name"], order_by="name")
    manufacturers = frappe.get_all("Distribution Organization", 
                                 filters={"organization_type": ["in", ["Manufacturer", "Supplier"]]},
                                 fields=["name", "organization_name"], 
                                 order_by="organization_name")
    users = frappe.get_all("User", 
                          filters={"enabled": 1, "user_type": "System User"},
                          fields=["name", "full_name"],
                          order_by="full_name")
    
    return {
        "countries": [d.name for d in countries],
        "manufacturers": [{"value": d.name, "label": d.organization_name} for d in manufacturers],
        "users": [{"value": d.name, "label": d.full_name} for d in users],
        "compliance_statuses": ["Compliant", "Pending Review", "Non-Compliant", "Partially Compliant", "Expired", "Not Applicable"],
        "risk_levels": ["Low", "Medium", "High", "Critical"]
    }

File: sysmayal7/sysmayal2/report/compliance_status_report/compliance_status_report.js
/**
 * Compliance Status Report - Client Side Configuration
 */

frappe.query_reports["Compliance Status Report"] = {
    "filters": [
        {
            "fieldname": "country",
            "label": __("Country"),
            "fieldtype": "Link",
            "options": "Country",
            "width": "100px"
        },
        {
            "fieldname": "compliance_status",
            "label": __("Compliance Status"),
            "fieldtype": "Select",
            "options": "\nCompliant\nPending Review\nNon-Compliant\nPartially Compliant\nExpired\nNot Applicable",
            "width": "100px"
        },
        {
            "fieldname": "risk_level",
            "label": __("Risk Level"),
            "fieldtype": "Select",
            "options": "\nLow\nMedium\nHigh\nCritical",
            "width": "100px"
        },
        {
            "fieldname": "manufacturer",
            "label": __("Manufacturer"),
            "fieldtype": "Link",
            "options": "Distribution Organization",
            "get_query": function() {
                return {
                    "filters": {
                        "organization_type": ["in", ["Manufacturer", "Supplier"]]
                    }
                };
            },
            "width": "100px"
        },
        {
            "fieldname": "responsible_person",
            "label": __("Responsible Person"),
            "fieldtype": "Link",
            "options": "User",
            "width": "100px"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "width": "80px"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "width": "80px"
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        // Color coding for compliance status
        if (column.fieldname == "compliance_status") {
            if (value == "Compliant") {
                value = `<span style="color: #28a745; font-weight: bold;">${value}</span>`;
            } else if (value == "Non-Compliant" || value == "Expired") {
                value = `<span style="color: #dc3545; font-weight: bold;">${value}</span>`;
            } else if (value == "Pending Review" || value == "Partially Compliant") {
                value = `<span style="color: #ffc107; font-weight: bold;">${value}</span>`;
            }
        }

        // Color coding for risk level
        if (column.fieldname == "risk_level") {
            if (value == "Critical") {
                value = `<span style="color: #dc3545; font-weight: bold;">${value}</span>`;
            } else if (value == "High") {
                value = `<span style="color: #fd7e14; font-weight: bold;">${value}</span>`;
            } else if (value == "Medium") {
                value = `<span style="color: #ffc107;">${value}</span>`;
            } else if (value == "Low") {
                value = `<span style="color: #28a745;">${value}</span>`;
            }
        }

        // Highlight expiring products
        if (column.fieldname == "days_to_expiry") {
            if (data && data.days_to_expiry !== null) {
                if (data.days_to_expiry < 0) {
                    value = `<span style="color: #dc3545; font-weight: bold;">${value} (Expired)</span>`;
                } else if (data.days_to_expiry <= 30) {
                    value = `<span style="color: #fd7e14; font-weight: bold;">${value}</span>`;
                } else if (data.days_to_expiry <= 90) {
                    value = `<span style="color: #ffc107;">${value}</span>`;
                }
            }
        }

        // Highlight overdue reviews
        if (column.fieldname == "days_to_review") {
            if (data && data.days_to_review !== null) {
                if (data.days_to_review < 0) {
                    value = `<span style="color: #dc3545; font-weight: bold;">${value} (Overdue)</span>`;
                } else if (data.days_to_review <= 7) {
                    value = `<span style="color: #fd7e14; font-weight: bold;">${value}</span>`;
                } else if (data.days_to_review <= 30) {
                    value = `<span style="color: #ffc107;">${value}</span>`;
                }
            }
        }

        // Progress bar for compliance percentage
        if (column.fieldname == "compliance_percentage") {
            if (data && data.compliance_percentage !== null) {
                const percentage = data.compliance_percentage;
                let color = "#28a745"; // Green for high compliance
                
                if (percentage < 50) {
                    color = "#dc3545"; // Red for low compliance
                } else if (percentage < 80) {
                    color = "#ffc107"; // Yellow for medium compliance
                }
                
                value = `
                    <div style="display: flex; align-items: center;">
                        <div style="
                            width: 60px; 
                            height: 8px; 
                            background: #f5f5f5; 
                            border-radius: 4px; 
                            margin-right: 8px; 
                            overflow: hidden;
                        ">
                            <div style="
                                width: ${percentage}%; 
                                height: 100%; 
                                background: ${color}; 
                                transition: width 0.3s ease;
                            "></div>
                        </div>
                        <span>${percentage}%</span>
                    </div>
                `;
            }
        }

        return value;
    },

    "onload": function(report) {
        // Add custom buttons
        report.page.add_inner_button(__("Export Compliance Summary"), function() {
            export_compliance_summary(report);
        });

        report.page.add_inner_button(__("Send Alerts"), function() {
            send_compliance_alerts(report);
        });

        report.page.add_inner_button(__("Update Status"), function() {
            bulk_update_compliance_status(report);
        });
    }
};

function export_compliance_summary(report) {
    // Export compliance summary
    const data = report.data;
    if (!data || data.length === 0) {
        frappe.msgprint(__("No data to export"));
        return;
    }

    // Generate summary data
    const summary = generate_summary_data(data);
    
    // Create and download CSV
    const csv_data = convert_to_csv(summary);
    download_csv(csv_data, "compliance_summary.csv");
}

function send_compliance_alerts(report) {
    const data = report.data;
    if (!data || data.length === 0) {
        frappe.msgprint(__("No data available"));
        return;
    }

    // Find products that need alerts
    const alerts_needed = data.filter(row => {
        return (row.days_to_expiry && row.days_to_expiry <= 30 && row.days_to_expiry > 0) ||
               (row.days_to_review && row.days_to_review <= 7 && row.days_to_review > 0) ||
               row.compliance_status === "Non-Compliant";
    });

    if (alerts_needed.length === 0) {
        frappe.msgprint(__("No alerts needed at this time"));
        return;
    }

    frappe.confirm(
        __("Send compliance alerts for {0} products?", [alerts_needed.length]),
        function() {
            frappe.call({
                method: "sysmayal.sysmayal_module.report.compliance_status_report.compliance_status_report.send_compliance_alerts",
                args: {
                    products: alerts_needed
                },
                callback: function(r) {
                    if (r.message) {
                        frappe.msgprint(__("Alerts sent successfully"));
                    }
                }
            });
        }
    );
}

function bulk_update_compliance_status(report) {
    const selected_rows = report.datatable.rowmanager.getCheckedRows();
    
    if (selected_rows.length === 0) {
        frappe.msgprint(__("Please select products to update"));
        return;
    }

    const dialog = new frappe.ui.Dialog({
        title: __("Update Compliance Status"),
        fields: [
            {
                fieldname: "new_status",
                label: __("New Status"),
                fieldtype: "Select",
                options: "Compliant\nPending Review\nNon-Compliant\nPartially Compliant\nExpired\nNot Applicable",
                reqd: 1
            },
            {
                fieldname: "notes",
                label: __("Notes"),
                fieldtype: "Text",
                description: __("Optional notes about the status change")
            }
        ],
        primary_action_label: __("Update"),
        primary_action: function(values) {
            const product_names = selected_rows.map(idx => report.data[idx].name);
            
            frappe.call({
                method: "sysmayal.sysmayal_module.doctype.product_compliance.product_compliance.bulk_update_compliance_status",
                args: {
                    product_names: product_names,
                    new_status: values.new_status,
                    notes: values.notes
                },
                callback: function(r) {
                    if (r.message) {
                        frappe.msgprint(r.message);
                        report.refresh();
                        dialog.hide();
                    }
                }
            });
        }
    });

    dialog.show();
}

function generate_summary_data(data) {
    // Generate summary statistics
    const total = data.length;
    const compliant = data.filter(row => row.compliance_status === "Compliant").length;
    const non_compliant = data.filter(row => row.compliance_status === "Non-Compliant").length;
    const expired = data.filter(row => row.compliance_status === "Expired").length;
    const expiring_soon = data.filter(row => row.days_to_expiry && row.days_to_expiry <= 30 && row.days_to_expiry > 0).length;

    return [
        ["Metric", "Count", "Percentage"],
        ["Total Products", total, "100%"],
        ["Compliant", compliant, `${(compliant/total*100).toFixed(1)}%`],
        ["Non-Compliant", non_compliant, `${(non_compliant/total*100).toFixed(1)}%`],
        ["Expired", expired, `${(expired/total*100).toFixed(1)}%`],
        ["Expiring Soon", expiring_soon, `${(expiring_soon/total*100).toFixed(1)}%`]
    ];
}

function convert_to_csv(data) {
    return data.map(row => row.join(",")).join("\n");
}

function download_csv(csv_data, filename) {
    const blob = new Blob([csv_data], { type: "text/csv;charset=utf-8;" });
    const link = document.createElement("a");
    
    if (link.download !== undefined) {
        const url = URL.createObjectURL(blob);
        link.setAttribute("href", url);
        link.setAttribute("download", filename);
        link.style.visibility = "hidden";
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
    }
}
